Throws an exception when trying to parse complex SQL statements with UNION and SELECT

Trying to compile complex SQL queries such as

```SQL
getFiltered:
SELECT IFNULL(t.filtered, :defaultValue)
FROM (
    SELECT filtered
    FROM filtered
    WHERE bufferId = :bufferId
    AND accountId = :accountId
    UNION
    SELECT NULL
    ORDER BY filtered
    DESC LIMIT 1
) t;
```

fails with an exception:
```
Caused by: java.lang.IllegalStateException: Failed to compile SqliteCompoundSelectStmtImpl(COMPOUND_SELECT_STMT): [] :
SELECT IFNULL(t.filtered, :defaultValue)
FROM (
    SELECT filtered
    FROM filtered
    WHERE bufferId = :bufferId
    AND accountId = :accountId
    UNION
    SELECT NULL
    ORDER BY filtered
    DESC LIMIT 1
) t
        at com.squareup.sqldelight.core.compiler.SqlDelightCompilerKt.tryWithElement(SqlDelightCompiler.kt:119)
        at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator.generateType(QueriesTypeGenerator.kt:59)
        at com.squareup.sqldelight.core.compiler.SqlDelightCompiler.writeQueriesType$sqldelight_compiler(SqlDelightCompiler.kt:80)
        at com.squareup.sqldelight.core.compiler.SqlDelightCompiler.compile(SqlDelightCompiler.kt:36)
        at com.squareup.sqldelight.core.SqlDelightEnvironment$generateSqlDelightFiles$2.invoke(SqlDelightEnvironment.kt:112)
        at com.squareup.sqldelight.core.SqlDelightEnvironment$generateSqlDelightFiles$2.invoke(SqlDelightEnvironment.kt:53)
        at com.alecstrong.sqlite.psi.core.SqliteCoreEnvironment$forSourceFiles$1.processFile(SqliteCoreEnvironment.kt:69)
        at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:100)
        at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:97)
        at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContentUnderDirectory(SqliteCoreEnvironment.kt:97)
        at com.alecstrong.sqlite.psi.core.CoreFileIndex.iterateContent(SqliteCoreEnvironment.kt:91)
        at com.alecstrong.sqlite.psi.core.SqliteCoreEnvironment.forSourceFiles(SqliteCoreEnvironment.kt:67)
        at com.squareup.sqldelight.core.SqlDelightEnvironment.generateSqlDelightFiles(SqlDelightEnvironment.kt:109)
        at com.squareup.sqldelight.gradle.SqlDelightTask.generateSqlDelightFiles(SqlDelightTask.kt:50)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:801)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:768)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:300)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:292)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:174)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:90)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        ... 31 more
Caused by: java.lang.AssertionError: Cannot bind unknown types or null
        at com.squareup.sqldelight.core.lang.IntermediateType$SqliteType.prepareStatementBinder(IntermediateType.kt:149)
        at com.squareup.sqldelight.core.lang.IntermediateType.preparedStatementBinder(IntermediateType.kt:89)
        at com.squareup.sqldelight.core.compiler.QueryGenerator.executeBlock(QueryGenerator.kt:105)
        at com.squareup.sqldelight.core.compiler.SelectQueryGenerator.querySubtype(SelectQueryGenerator.kt:197)
        at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator$generateType$$inlined$forEach$lambda$1.invoke(QueriesTypeGenerator.kt:70)
        at com.squareup.sqldelight.core.compiler.QueriesTypeGenerator$generateType$$inlined$forEach$lambda$1.invoke(QueriesTypeGenerator.kt:19)
        at com.squareup.sqldelight.core.compiler.SqlDelightCompilerKt.tryWithElement(SqlDelightCompiler.kt:117)
        ... 58 more
```

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/IntermediateType.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/IntermediateType.kt
@@ -132,7 +132,7 @@ internal data class IntermediateType(
   }
 
   enum class SqliteType(val javaType: TypeName) {
-    ARGUMENT(ANY),
+    ARGUMENT(ANY.copy(nullable = true)),
     NULL(Nothing::class.asClassName().copy(nullable = true)),
     INTEGER(LONG),
     REAL(DOUBLE),

--- a/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/util/Arguments.kt
+++ b/sqldelight-compiler/src/main/kotlin/com/squareup/sqldelight/core/lang/util/Arguments.kt
@@ -111,7 +111,7 @@ private fun SqliteExpr.argumentType(argument: SqliteExpr): IntermediateType {
     }
 
     is SqliteFunctionExpr -> {
-      return IntermediateType(ARGUMENT)
+      return type()
     }
     else -> throw AssertionError()
   }

--- a/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/BindableQueryTest.kt
+++ b/sqldelight-compiler/src/test/kotlin/com/squareup/sqldelight/core/queries/BindableQueryTest.kt
@@ -7,7 +7,9 @@ import com.intellij.psi.util.PsiTreeUtil
 import com.squareup.kotlinpoet.LONG
 import com.squareup.kotlinpoet.asClassName
 import com.squareup.sqldelight.core.lang.IntermediateType
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.ARGUMENT
 import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.INTEGER
+import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.NULL
 import com.squareup.sqldelight.core.lang.IntermediateType.SqliteType.TEXT
 import com.squareup.sqldelight.core.lang.psi.ColumnDefMixin
 import com.squareup.sqldelight.test.util.FixtureCompiler
@@ -109,7 +111,7 @@ class BindableQueryTest {
   }
 
   @Test fun `auto-generated parameter from earlier in query has conflicting name with user-specified name`() {
-      val file = FixtureCompiler.parseSql("""
+    val file = FixtureCompiler.parseSql("""
       |CREATE TABLE data (
       |  _id INTEGER NOT NULL PRIMARY KEY,
       |  value TEXT AS kotlin.collections.List
@@ -128,6 +130,39 @@ class BindableQueryTest {
     assertThat(update.arguments.map { it.index to it.type }).containsExactly(
         1 to IntermediateType(TEXT, List::class.asClassName().copy(nullable = true), createTable.column(1), "value_", args[0]),
         2 to IntermediateType(TEXT, List::class.asClassName().copy(nullable = true), createTable.column(1), "value", args[1])
+    )
+  }
+
+  @Test fun `argument in ifnull statement`() {
+    val file = FixtureCompiler.parseSql("""
+      |CREATE TABLE filtered (
+      |  filtered TEXT,
+      |  bufferId INTEGER,
+      |  accountId INTEGER
+      |);
+      |
+      |updateStuff:
+      |SELECT IFNULL(t.filtered, :defaultValue)
+      |FROM (
+      |    SELECT filtered
+      |    FROM filtered
+      |    WHERE bufferId = :bufferId
+      |    AND accountId = :accountId
+      |    UNION
+      |    SELECT NULL
+      |    ORDER BY filtered
+      |    DESC LIMIT 1
+      |) t;
+      """.trimMargin(), tempFolder)
+
+    val createTable = file.sqliteStatements().mapNotNull { it.statement.createTableStmt }.first()
+    val select = file.namedQueries.first()
+    val args = PsiTreeUtil.findChildrenOfType(file, SqliteBindExpr::class.java).toTypedArray()
+
+    assertThat(select.arguments.map { it.index to it.type }).containsExactly(
+        1 to IntermediateType(TEXT, String::class.asClassName().copy(nullable = true), null, "defaultValue", args[0]),
+        2 to IntermediateType(INTEGER, Long::class.asClassName().copy(nullable = true), createTable.column(1), "bufferId", args[1]),
+        3 to IntermediateType(INTEGER, Long::class.asClassName().copy(nullable = true), createTable.column(2), "accountId", args[2])
     )
   }
 

